scripts/utils/options.ts

```diff
@@ -0,0 +1,245 @@
+// as `StringArrayOptionMatch` above. In both cases, the field should only ever be set to true
+export type OptionValue<TOption extends Option> = TOption extends { required: boolean }
+  ? string
+  : MaybeOptionValue<TOption>;
+
+export type OptionSpecifier = Record<OptionId, Option>;
+export type MaybeOptionValues<TOptions extends OptionSpecifier> = {
+  [TKey in keyof TOptions]: MaybeOptionValue<TOptions[TKey]>;
+};
+
+export type OptionValues<TOptions extends OptionSpecifier> = {
+  [TKey in keyof TOptions]: OptionValue<TOptions[TKey]>;
+};
+
+export function isStringOption(option: Option): option is StringOption {
+  return 'values' in option && !('multiple' in option);
+}
+
+export function isBooleanOption(option: Option): option is BooleanOption {
+  return !('values' in option);
+}
+
+export function isStringArrayOption(option: Option): option is StringArrayOption {
+  return 'values' in option && 'multiple' in option;
+}
+
+function shortFlag(key: OptionId, option: Option) {
+  const inverse = isBooleanOption(option) && option.inverse;
+  const defaultShortFlag = inverse ? key.substring(0, 1).toUpperCase() : key.substring(0, 1);
+  const short = option.shortFlag || defaultShortFlag;
+  if (short.length !== 1) {
+    throw new Error(
+      `Invalid shortFlag for ${key}: '${short}', needs to be a single character (e.g. 's')`
+    );
+  }
+  return short;
+}
+
+function longFlag(key: OptionId, option: Option) {
+  const inverse = isBooleanOption(option) && option.inverse;
+  return inverse ? `no-${kebabCase(key)}` : kebabCase(key);
+}
+
+function optionFlags(key: OptionId, option: Option) {
+  const base = `-${shortFlag(key, option)}, --${longFlag(key, option)}`;
+  if (isStringOption(option) || isStringArrayOption(option)) {
+    return `${base} <${key}>`;
+  }
+  return base;
+}
+
+export function getOptions<TOptions extends OptionSpecifier>(
+  command: Command,
+  options: TOptions,
+  argv: string[]
+): MaybeOptionValues<TOptions> {
+  Object.entries(options)
+    .reduce((acc, [key, option]) => {
+      const flags = optionFlags(key, option);
+
+      if (isBooleanOption(option)) return acc.option(flags, option.description, !!option.inverse);
+
+      const checkStringValue = (raw: string) => {
+        if (!option.values.includes(raw))
+          throw new Error(`Unexpected value '${raw}' for option '${key}'`);
+        return raw;
+      };
+
+      if (isStringOption(option))
+        return acc.option(flags, option.description, (raw) => checkStringValue(raw));
+
+      if (isStringArrayOption(option)) {
+        return acc.option(
+          flags,
+          option.description,
+          (raw, values) => [...values, checkStringValue(raw)],
+          []
+        );
+      }
+
+      throw new Error(`Unexpected option type '${key}'`);
+    }, command)
+    .parse(argv);
+
+  // Note the code above guarantees the types as they come in, so we cast here.
+  // Not sure there is an easier way to do this
```

Comment:
There is, but we need to upgrade `commander`. We are currently using v6 while there's already v9 out there.
v8 released a feature to [allow generics to be passed](https://github.com/tj/commander.js/pull/1539) so we can upgrade the lib to make it work

Reply:
OK, I'll do that in a separate PR?

Reply:
Actually I wonder if there's a heap of benefit to doing this. I mean we may as well update but it's just a different type of cast ultimately isn't it?